src/cpu/x64/jit_uni_reorder.cpp

```diff
@@ -1281,12 +1618,164 @@ struct jit_uni_reorder_t : public primitive_t {
+        }
+
+        tr::prb_t prb_;
+    };
+
+    jit_blk_reorder_t(const pd_t *apd) : primitive_t(apd) {
+        kernel_ = utils::make_unique<tr::jit_single_blk_kernel>(pd()->prb_);
+    }
+
+    inline int n(int d) const {
+        assert(d < pd()->prb_.ndims);
+        return (int)pd()->prb_.nodes[d].n;
+    }
+    inline int is(int d) const {
+        assert(d < pd()->prb_.ndims);
+        return (int)pd()->prb_.nodes[d].is;
+    }
+    inline int os(int d) const {
+        assert(d < pd()->prb_.ndims);
+        return (int)pd()->prb_.nodes[d].os;
+    }
+    inline int ss(int d) const {
+        assert(d < pd()->prb_.ndims);
+        return (int)pd()->prb_.nodes[d].ss;
+    }
+
+    status_t execute(const exec_ctx_t &ctx) const override {
+        auto in = CTX_IN_MEM(const char *, DNNL_ARG_FROM);
+        auto out = CTX_OUT_MEM(char *, DNNL_ARG_TO);
+        DEFINE_SCALES_BUFFER(scales);
+
+        // kernel handle 2-dimension tiles
+        // Which means there is a tail
+        auto &prb = this->pd()->prb_;
+        ptrdiff_t BH = 1;
+        for (int i = 2; i < prb.ndims; ++i) {
+            BH *= prb.nodes[i].n;
+        }
+
+        auto block_sz = n(0);
+        auto FL = (n(1) + block_sz - 1) / block_sz;
+        auto bh_stride = BH == 1 ? 0 /* don't care, and no is(2) */ : is(2);
+
+        auto itype_sz = data_type_size(pd()->prb_.itype);
+        auto otype_sz = data_type_size(pd()->prb_.otype);
+
+        if (kernel_->has_tail()) {
+            const size_t work_amount = (size_t)BH * FL;
+            int nthr = adjust_num_threads(dnnl_get_max_threads(), work_amount);
+            if (nthr)
+                parallel(nthr, [&](int ithr, int nthr) {
+                    kernel_->set_mask();
+                    for_nd(ithr, nthr, BH, FL, [&](dim_t bh, dim_t fl) {
+                        auto fl_b = fl * block_sz;
+                        auto bh_b = bh_stride * bh;
+                        auto *i = in + (bh_b + fl_b * is(1)) * itype_sz;
+                        auto *o = out + (bh_b + fl_b * os(1)) * otype_sz;
+                        (*kernel_)(i, o, nullptr, n(1) - fl_b < block_sz);
+                    });
+                    kernel_->clean_avx();
+                });
+        } else {
+            const size_t work_amount = (size_t)BH * FL;
+            int nthr = adjust_num_threads(dnnl_get_max_threads(), work_amount);
+            if (nthr)
+                parallel(nthr, [&](int ithr, int nthr) {
+                    for_nd(ithr, nthr, BH, FL, [&](dim_t bh, dim_t fl) {
+                        auto fl_b = fl * block_sz;
+                        auto bh_b = bh_stride * bh;
+                        auto *i = in + (bh_b + fl_b * is(1)) * itype_sz;
+                        auto *o = out + (bh_b + fl_b * os(1)) * otype_sz;
+                        (*kernel_)(i, o, nullptr);
+                    });
+                    kernel_->clean_avx();
+                });
+        }
```

Comment:
the code looks pretty much the same. can we simplify and join the if-else sections?